File: dart/lib/src/redundant.dart
// L3 — aggregate several unreliable links into ONE reliable, ordered, never-interrupted duplex byte
// stream, the Dart mirror of the peers. Every byte is written to all live links, delivered once in
// order by absolute offset, acknowledged cumulatively, and a dropped link reconnects and replays the
// unacknowledged tail. write() completes when the send window has room.

import 'dart:async';
import 'dart:math';
import 'dart:typed_data';

import 'frames.dart';
import 'link.dart';

/// Reported to [RedundantOptions.onLinkState] the moment a line's up/down state changes — the edge a
/// log wants. [up] is true on (re)connect, false on drop; [reason] is the drop cause (empty on up);
/// [durationMs] is how long the line spent in the state it just left.
class LinkState {
  final int index;
  final bool up;
  final String reason;
  final int durationMs;
  LinkState(this.index, this.up, this.reason, this.durationMs);
}

/// A point-in-time snapshot of one line — the level a status view wants. [state] is "up",
/// "connecting" (never yet up) or "down" (was up, now reconnecting). [lastByteMs] is when a frame
/// last arrived (0 if never). [reconnects] counts recoveries after a drop; [reason] is the last drop
/// cause.
class LinkStat {
  final int index;
  final String state;
  final int lastByteMs;
  final int reconnects;
  final String reason;
  LinkStat(
      this.index, this.state, this.lastByteMs, this.reconnects, this.reason);
}

class RedundantOptions {
  final List<String> urls;
  final int window;
  final Duration pingInterval;
  final Duration deadAfter;
  final Duration ackInterval;
  final Duration reconnectDelay;
  final Duration maxDelay;
  final WsConnect? connect;

  /// Called once per line up/down transition (never on the hot path). Edge-triggered: a repeatedly
  /// failing reconnect does not re-fire it, but the latest reason shows up in [RedundantStream.stats].
  final void Function(LinkState)? onLinkState;
  RedundantOptions({
    required this.urls,
    this.window = 4 << 20,
    this.pingInterval = const Duration(seconds: 5),
    this.deadAfter = const Duration(seconds: 15),
    this.ackInterval = const Duration(milliseconds: 50),
    this.reconnectDelay = const Duration(milliseconds: 200),
    this.maxDelay = const Duration(seconds: 5),
    this.connect,
    this.onLinkState,
  });
}

class _Slot {
  Link? link;
  int lastSeen;
  int backoffMs;
  Timer? reconnectTimer;
  String state = 'connecting';
  int sinceMs;
  int reconnects = 0;
  String reason = '';
  String reapReason = '';
  bool rejected = false; // origin refused this link (REJECT) — never retry it
  _Slot(this.lastSeen, this.backoffMs) : sinceMs = lastSeen;
}

class RedundantStream {
  final RedundantOptions opt;
  final Uint8List connId;
  final List<_Slot> _slots;

  Uint8List _sendBuf =
      Uint8List(0); // unacknowledged bytes: [ackedOffset, sendOffset)
  int _sendOffset = 0;
  int _ackedOffset = 0;
  int _recvOffset = 0;
  bool _ackScheduled = false;
  final List<Completer<void>> _windowWaiters = [];
  Timer? _pingTimer;
  bool _closed = false;
  bool _opened = false;
  final Completer<void> _openCompleter = Completer<void>();

  void Function(Uint8List) onDeliver = (_) {};

  /// Called when the stream closes; [err] is set when it closed with a cause (e.g. all links
  /// rejected). The mux above uses it to fail its streams so reads/writes don't hang.
  void Function(Object? err) onClose = (_) {};

  Object? _closeErr;

  RedundantStream(this.opt)
      : connId = _randomConnId(),
        _slots = List.generate(
          opt.urls.length,
          (_) => _Slot(DateTime.now().millisecondsSinceEpoch,
              opt.reconnectDelay.inMilliseconds),
        );

  static Uint8List _randomConnId() {
    final r = Random.secure();
    return Uint8List.fromList(List.generate(16, (_) => r.nextInt(256)));
  }

  int get _nowMs => DateTime.now().millisecondsSinceEpoch;

  Future<void> dial() async {
    for (var i = 0; i < _slots.length; i++) {
      _connect(i);
    }
    _pingTimer = Timer.periodic(opt.pingInterval, (_) => _tick());
    await _openCompleter.future;
  }

  // _markUp records line i as connected, firing onLinkState on a real transition (edge-triggered).
  void _markUp(int i) {
    final slot = _slots[i];
    if (slot.state == 'up') return;
    final was = slot.state;
    final dur = _nowMs - slot.sinceMs;
    slot.state = 'up';
    slot.sinceMs = _nowMs;
    if (was == 'down') slot.reconnects++;
    opt.onLinkState?.call(LinkState(i, true, '', dur));
  }

  // _markDown records line i as down with a reason. Always updates the last reason; fires
  // onLinkState only on a real transition, so repeated failed reconnects don't spam.
  void _markDown(int i, String reason) {
    final slot = _slots[i];
    slot.reason = reason;
    if (slot.state == 'down') return;
    final dur = _nowMs - slot.sinceMs;
    slot.state = 'down';
    slot.sinceMs = _nowMs;
    opt.onLinkState?.call(LinkState(i, false, reason, dur));
  }

  /// A snapshot of every line's current health. A status view reads this; a log uses onLinkState.
  List<LinkStat> stats() => List.generate(
        _slots.length,
        (i) => LinkStat(i, _slots[i].state, _slots[i].lastSeen,
            _slots[i].reconnects, _slots[i].reason),
      );

  void _connect(int i) {
    if (_closed) return;
    final slot = _slots[i];
    openWebSocketLink(
      opt.urls[i],
      connId,
      i,
      LinkHandlers(
        onOpen: () {
          slot.backoffMs = opt.reconnectDelay.inMilliseconds;
          slot.lastSeen = _nowMs;
          slot.reapReason = '';
          _replay(slot);
          _markUp(i);
          if (!_opened) {
            _opened = true;
            if (!_openCompleter.isCompleted) _openCompleter.complete();
          }
        },
        onFrame: (frame) {
          slot.lastSeen = _nowMs;
          _onFrame(frame, slot, i);
        },
        onClose: () {
          slot.link = null;
          if (slot.rejected)
            return; // refused by origin; already down, never retry
          _markDown(
              i,
              slot.reapReason.isNotEmpty
                  ? slot.reapReason
                  : 'connection closed');
          slot.reapReason = '';
          _scheduleReconnect(i);
        },
      ),
      connect: opt.connect,
    ).then((link) {
      if (_closed) {
        link.close();
      } else {
        slot.link = link;
      }
    }).catchError((Object e) {
      _markDown(i, 'connect failed: $e');
      _scheduleReconnect(i);
    });
  }

  void _scheduleReconnect(int i) {
    if (_closed) return;
    final slot = _slots[i];
    if (slot.reconnectTimer != null) return;
    slot.reconnectTimer = Timer(Duration(milliseconds: slot.backoffMs), () {
      slot.reconnectTimer = null;
      slot.backoffMs = min(slot.backoffMs * 2, opt.maxDelay.inMilliseconds);
      _connect(i);
    });
  }

  void _onFrame(Frame frame, _Slot slot, int i) {
    switch (frame.type) {
      case frameReject:
        // The origin refused this link and said why. Mark it down, never retry it, and if every
        // link is refused, close the stream with the reason so reads/writes fail fast.
        slot.rejected = true;
        _markDown(i, 'origin rejected: ${frame.reason}');
        slot.link?.close();
        slot.link = null;
        if (_slots.every((s) => s.rejected)) {
          _closeWithError(StateError(
              'multipath: all links rejected by origin: ${frame.reason}'));
        }
        return;
      case frameData:
        final offset = frame.offset;
        final payload = frame.payload!;
        final end = offset + payload.length;
        if (end <= _recvOffset) return; // fully seen
        if (offset > _recvOffset) return; // a gap cannot happen on a live link
        final start = _recvOffset - offset;
        _recvOffset = end;
        onDeliver(Uint8List.sublistView(payload, start));
        _scheduleAck();
        break;
      case frameAck:
        final cum = frame.offset;
        if (cum > _ackedOffset) {
          final drop = min(cum - _ackedOffset, _sendBuf.length);
          _sendBuf = Uint8List.sublistView(_sendBuf, drop);
          _ackedOffset = cum;
          _wakeWriters();
        }
        break;
      case framePing:
        slot.link?.send(encodeNonce(framePong, frame.nonce));
        break;
      case framePong:
        break;
    }
  }

  Future<void> write(Uint8List bytes) async {
    if (_closed)
      throw _closeErr ?? StateError('multipath: redundant stream closed');
    var sent = 0;
    while (sent < bytes.length) {
      await _awaitWindow();
      if (_closed)
        throw _closeErr ?? StateError('multipath: redundant stream closed');
      final room = opt.window - _inFlight;
      final n = min(bytes.length - sent, room);
      final chunk = Uint8List.sublistView(bytes, sent, sent + n);
      _append(chunk);
      var off = _sendOffset - n;
      for (var p = 0; p < n; p += maxSegment) {
        final seg = Uint8List.sublistView(chunk, p, min(p + maxSegment, n));
        final framed = encodeData(off, seg);
        for (final slot in _slots) {
          slot.link?.send(framed);
        }
        off += seg.length;
      }
      sent += n;
    }
  }

  void _append(Uint8List chunk) {
    final next = Uint8List(_sendBuf.length + chunk.length);
    next.setRange(0, _sendBuf.length, _sendBuf);
    next.setRange(_sendBuf.length, next.length, chunk);
    _sendBuf = next;
    _sendOffset += chunk.length;
  }

  void _replay(_Slot slot) {
    if (_sendBuf.isEmpty) return;
    var off = _ackedOffset;
    for (var p = 0; p < _sendBuf.length; p += maxSegment) {
      final seg = Uint8List.sublistView(
          _sendBuf, p, min(p + maxSegment, _sendBuf.length));
      slot.link?.send(encodeData(off, seg));
      off += seg.length;
    }
  }

  int get _inFlight => _sendOffset - _ackedOffset;

  Future<void> _awaitWindow() {
    if (_inFlight < opt.window) return Future.value();
    final c = Completer<void>();
    _windowWaiters.add(c);
    return c.future;
  }

  void _wakeWriters() {
    final waiters = List<Completer<void>>.from(_windowWaiters);
    _windowWaiters.clear();
    for (final c in waiters) {
      if (!c.isCompleted) c.complete();
    }
  }

  void _scheduleAck() {
    if (_ackScheduled) return;
    _ackScheduled = true;
    Timer(opt.ackInterval, () {
      _ackScheduled = false;
      final ack = encodeAck(_recvOffset);
      for (final slot in _slots) {
        slot.link?.send(ack);
      }
    });
  }

  void _tick() {
    final now = _nowMs;
    for (var i = 0; i < _slots.length; i++) {
      final slot = _slots[i];
      final link = slot.link;
      if (link == null) continue;
      if (now - slot.lastSeen > opt.deadAfter.inMilliseconds) {
        slot.link = null;
        // close() drives the link's onClose, which marks it down; stamp the real cause first.
        slot.reapReason = 'no data for ${opt.deadAfter.inMilliseconds}ms';
        link.close();
        _scheduleReconnect(i);
        continue;
      }
      link.send(encodeNonce(framePing, i));
    }
  }

  // _closeWithError records a cause then closes, so the mux above (via onClose) fails its streams
  // with it and blocked/subsequent reads and writes surface the reason instead of hanging.
  void _closeWithError(Object err) {
    if (_closed) return;
    _closeErr = err;
    close();
  }

  void close() {
    if (_closed) return;
    _closed = true;
    _pingTimer?.cancel();
    for (final slot in _slots) {
      slot.reconnectTimer?.cancel();
      slot.link?.close();
    }
    _wakeWriters();
    if (!_openCompleter.isCompleted) _openCompleter.complete();
    onClose(_closeErr);
  }
}

File: ts/src/redundant.ts
// L3 — aggregate several unreliable links into ONE reliable, ordered, never-interrupted duplex byte
// stream, the browser's async mirror of the Go peer (redundant.go). The model is "redundant, not
// striped": every byte is written to ALL live links, the receiver delivers each byte once in order
// by absolute offset, and a link that dies reconnects and replays the unacknowledged tail behind the
// scenes. As long as one link is healthy the logical stream never stalls and never breaks.
//
// Where Go blocks a goroutine, the browser awaits a promise: write() resolves when the send window
// has room, and delivered bytes are handed up through onDeliver. connID is the client's, minted once
// with the platform CSPRNG and carried in every link's HELLO so the origin groups them.

import { FrameType, MAX_SEGMENT, encodeAck, encodeData, encodeNonce } from './frames.js';
import { Link, openWebSocketLink } from './link.js';

export interface RedundantOptions {
  urls: string[]; // one per line; a link is opened (and reconnected) to each
  window?: number; // max unacknowledged bytes before write() awaits ACKs; default 4MiB
  pingIntervalMs?: number; // default 5000
  deadAfterMs?: number; // default 15000; a link silent this long is reconnected
  ackIntervalMs?: number; // default 50
  reconnectDelayMs?: number; // default 200, doubling to maxDelayMs
  maxDelayMs?: number; // default 5000
  wsCtor?: typeof WebSocket; // injected in tests
  now?: () => number; // injected in tests
}

interface LinkSlot {
  link: Link | null;
  lastSeen: number;
  backoff: number;
  reconnectTimer: ReturnType<typeof setTimeout> | null;
}

export class RedundantStream {
  readonly connID: Uint8Array;
  private opt: Required<Omit<RedundantOptions, 'wsCtor' | 'now'>>;
  private wsCtor: typeof WebSocket;
  private now: () => number;

  private slots: LinkSlot[];
  private sendBuf = new Uint8Array(0); // unacknowledged bytes: [ackedOffset, sendOffset)
  private sendOffset = 0n;
  private ackedOffset = 0n;
  private recvOffset = 0n;
  private ackScheduled = false;
  private windowWaiters: Array<() => void> = [];
  private pingTimer: ReturnType<typeof setInterval> | null = null;
  private closed = false;
  private opened = false;
  private onOpenResolve: (() => void) | null = null;
  private openPromise: Promise<void>;

  /** Delivered, in-order bytes. Set by the mux above before any traffic flows. */
  onDeliver: (bytes: Uint8Array) => void = () => {};

  constructor(opt: RedundantOptions) {
    this.opt = {
      urls: opt.urls,
      window: opt.window ?? 4 << 20,
      pingIntervalMs: opt.pingIntervalMs ?? 5000,
      deadAfterMs: opt.deadAfterMs ?? 15000,
      ackIntervalMs: opt.ackIntervalMs ?? 50,
      reconnectDelayMs: opt.reconnectDelayMs ?? 200,
      maxDelayMs: opt.maxDelayMs ?? 5000,
    };
    this.wsCtor = opt.wsCtor ?? WebSocket;
    this.now = opt.now ?? Date.now;
    this.connID = new Uint8Array(16);
    crypto.getRandomValues(this.connID);
    this.slots = this.opt.urls.map(() => ({
      link: null,
      lastSeen: this.now(),
      backoff: this.opt.reconnectDelayMs,
      reconnectTimer: null,
    }));
    this.openPromise = new Promise((resolve) => (this.onOpenResolve = resolve));
  }

  /** Opens the links and resolves once at least one is up. */
  async dial(): Promise<void> {
    for (let i = 0; i < this.slots.length; i++) this.connect(i);
    this.pingTimer = setInterval(() => this.tick(), this.opt.pingIntervalMs);
    await this.openPromise;
  }

  private connect(i: number): void {
    if (this.closed) return;
    const slot = this.slots[i];
    slot.link = openWebSocketLink(
      this.opt.urls[i],
      this.connID,
      i,
      {
        onOpen: () => {
          slot.backoff = this.opt.reconnectDelayMs;
          slot.lastSeen = this.now();
          this.replay(slot); // resume: resend everything still unacknowledged
          if (!this.opened) {
            this.opened = true;
            this.onOpenResolve?.();
          }
        },
        onFrame: (frame) => {
          slot.lastSeen = this.now();
          this.onFrame(frame, slot);
        },
        onClose: () => {
          slot.link = null;
          this.scheduleReconnect(i);
        },
      },
      this.wsCtor,
    );
  }

  private scheduleReconnect(i: number): void {
    if (this.closed) return;
    const slot = this.slots[i];
    if (slot.reconnectTimer !== null) return;
    slot.reconnectTimer = setTimeout(() => {
      slot.reconnectTimer = null;
      slot.backoff = Math.min(slot.backoff * 2, this.opt.maxDelayMs);
      this.connect(i);
    }, slot.backoff);
  }

  private onFrame(
    frame: {
      type: number;
      offset?: bigint;
      payload?: Uint8Array;
      cumulative?: bigint;
      nonce?: bigint;
    },
    slot: LinkSlot,
  ): void {
    switch (frame.type) {
      case FrameType.Data: {
        const offset = frame.offset!;
        const payload = frame.payload!;
        const end = offset + BigInt(payload.length);
        if (end <= this.recvOffset) return; // fully seen already
        if (offset > this.recvOffset) return; // a gap cannot happen on a live link; ignore
        const start = Number(this.recvOffset - offset);
        this.recvOffset = end;
        this.onDeliver(payload.subarray(start));
        this.scheduleAck();
        break;
      }
      case FrameType.Ack: {
        const cum = frame.cumulative!;
        if (cum > this.ackedOffset) {
          const drop = Number(cum - this.ackedOffset);
          this.sendBuf = this.sendBuf.subarray(Math.min(drop, this.sendBuf.length));
          this.ackedOffset = cum;
          this.wakeWriters();
        }
        break;
      }
      case FrameType.Ping:
        slot.link?.send(encodeNonce(FrameType.Pong, frame.nonce!));
        break;
      case FrameType.Pong:
        break;
    }
  }

  /** Writes bytes, awaiting the send window when it is full. */
  async write(bytes: Uint8Array): Promise<void> {
    if (this.closed) throw new Error('multipath: redundant stream closed');
    let sent = 0;
    while (sent < bytes.length) {
      await this.awaitWindow();
      if (this.closed) throw new Error('multipath: redundant stream closed');
      const room = this.opt.window - this.inFlight();
      const n = Math.min(bytes.length - sent, room);
      const chunk = bytes.subarray(sent, sent + n);
      this.append(chunk);
      // Fan the new bytes out to every live link, in maxSegment frames from their absolute offset.
      let off = this.sendOffset - BigInt(n);
      for (let p = 0; p < n; p += MAX_SEGMENT) {
        const seg = chunk.subarray(p, Math.min(p + MAX_SEGMENT, n));
        const framed = encodeData(off, seg);
        for (const slot of this.slots) slot.link?.send(framed);
        off += BigInt(seg.length);
      }
      sent += n;
    }
  }

  private append(chunk: Uint8Array): void {
    const next = new Uint8Array(this.sendBuf.length + chunk.length);
    next.set(this.sendBuf);
    next.set(chunk, this.sendBuf.length);
    this.sendBuf = next;
    this.sendOffset += BigInt(chunk.length);
  }

  // replay resends the whole unacknowledged tail to one freshly (re)connected link, so a dropped
  // link resumes with no gap. The receiver de-duplicates by offset, so re-sending is harmless.
  private replay(slot: LinkSlot): void {
    if (this.sendBuf.length === 0) return;
    let off = this.ackedOffset;
    for (let p = 0; p < this.sendBuf.length; p += MAX_SEGMENT) {
      const seg = this.sendBuf.subarray(p, Math.min(p + MAX_SEGMENT, this.sendBuf.length));
      slot.link?.send(encodeData(off, seg));
      off += BigInt(seg.length);
    }
  }

  private inFlight(): number {
    return Number(this.sendOffset - this.ackedOffset);
  }

  private awaitWindow(): Promise<void> {
    if (this.inFlight() < this.opt.window) return Promise.resolve();
    return new Promise((resolve) => this.windowWaiters.push(resolve));
  }

  private wakeWriters(): void {
    const waiters = this.windowWaiters;
    this.windowWaiters = [];
    for (const w of waiters) w();
  }

  private scheduleAck(): void {
    if (this.ackScheduled) return;
    this.ackScheduled = true;
    setTimeout(() => {
      this.ackScheduled = false;
      const ack = encodeAck(this.recvOffset);
      for (const slot of this.slots) slot.link?.send(ack);
    }, this.opt.ackIntervalMs);
  }

  private tick(): void {
    const now = this.now();
    for (let i = 0; i < this.slots.length; i++) {
      const slot = this.slots[i];
      if (slot.link === null) continue;
      if (now - slot.lastSeen > this.opt.deadAfterMs) {
        const dead = slot.link;
        slot.link = null;
        dead.close();
        this.scheduleReconnect(i);
        continue;
      }
      slot.link.send(encodeNonce(FrameType.Ping, BigInt(i)));
    }
  }

  close(): void {
    if (this.closed) return;
    this.closed = true;
    if (this.pingTimer !== null) clearInterval(this.pingTimer);
    for (const slot of this.slots) {
      if (slot.reconnectTimer !== null) clearTimeout(slot.reconnectTimer);
      slot.link?.close();
    }
    this.wakeWriters();
    this.onOpenResolve?.();
  }
}
